### ISSUE TITLE
CA2248 reports missing flags for generic enums

### ISSUE THREAD

**Issue by jnyrup** [dotnet/roslyn-analyzers#4181]
### Analyzer

**Diagnostic ID**: [CA2248](https://docs.microsoft.com/visualstudio/code-quality/CA2248)

### Describe the improvement

When using `HasFlag` for a generic enum type it cannot (<-- my assumption) know if the concrete enum is decorated with the `[Flags]` attribute.

### Describe suggestions on how to achieve the rule

```cs
public class StructEnums<T1>
        where T1 : struct, Enum
    {
        private readonly T1 value;

        public StructEnums(T1 value)
        {
            this.value = value;
        }

        public bool HaveFlag(T1 flag) =>
            value.HasFlag(flag); // CA2248: This call will always returns 'false' because the enum type 'T1' is not marked with 'FlagsAttribute'
    }
```

### Additional context

Greatly simplified example of how `StructEnums` is used
```cs
[Flags]
public enum SomeFlags
{
    A
}

[Flags]
public enum OtherFlags
{
    A
}

public static class Test
{
    public static void Main()
    {
        new StructEnums<SomeFlags>(SomeFlags.A).HaveFlag(SomeFlags.A);
        //new StructEnums<SomeFlags>(SomeFlags.A).HaveFlag(OtherFlags.A); // CS1503
        //new StructEnums<SomeFlags>(SomeFlags.A).HaveFlag((Enum)null); // CS1503

        //new StructEnums<Enum>(SomeFlags.A).HaveFlag(SomeFlags.A); // CS0453
        //new StructEnums<Enum>(SomeFlags.A).HaveFlag(OtherFlags.A); // CS0453
        //new StructEnums<Enum>(SomeFlags.A).HaveFlag(null); // CS0453
    }
}
```
